### PR TITLE
Testing To Do list: part 2

### PR DESCRIPTION
**Pull request by fredo509**
For this millstone we : 

- Use jest framework for testing.
- Create a test file(s) ([..].test.js) for a file(s) containing the following functions:
- a function for editing the task description.
- a function for updating an item's 'completed' status.
- the "Clear all completed" function.
- Use the mock storage object you have created in Part 1 of this project to mock the localStorage updates.
- Mock the HTML elements to test DOM manipulation functions.
- Make sure you group your tests using the describe() method.

### CODE REVIEW

File: src/task.test.js [hunk 1]
```diff
@@ -8,6 +9,13 @@ const dom = new JSDOM('<!DOCTYPE html><html><head></head><body></body></html>');
 global.window = dom.window;
 global.document = dom.window.document;
 
+beforeAll(() => {
+  Object.defineProperty(window, 'location', {
+    value: { reload: jest.fn() },
+    writable: true,
+  });
+});
+
```
Comment (adel-gu) [inline src/task.test.js:18]: - [x] For the testing `addTask` function you did great by expecting the `tasks` array in the localstorage to grow in length by one, however from the project requirement you suppose to test the dom-manipulation as well 👍🏼 this is noted from the screenshot below 💯 

Screenshot from 2023-04-06 22-08-32

### Suggestions 
Screenshot from 2023-04-06 22-13-00

- Here is your code after adding a new task you can consider testing the DOM by following:
```javascript
addTask(newTask);
renderTasks(); // Render the task into the dom
const newTaskDom = document.querySelector('li'); // select the first and only task from the dom
...

expect(newTaskDom).toBeTruthy();
```

File: src/task.test.js [hunk 2]
```diff
@@ -70,4 +78,74 @@ describe('deleteTask', () => {
...
+    /* eslint-enable global-require */
+    const tasksToCheck = JSON.parse(localStorage.getItem('tasksObj'));
+
+    ClearCompletedTasks(tasksToCheck);
+
+    const updatedTasks = JSON.parse(localStorage.getItem('tasksToCheck'));
+
+    expect(updatedTasks).toHaveLength(1);
+    expect(updatedTasks[0].completed).toEqual(false);
+  });
```
Comment (adel-gu) [inline src/task.test.js]: -   [x] Please consider testing  the mock DOM manipulation in the case of the function `ClearCompletedTasks` as well following the same example above 👍🏼

File: src/task.test.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import 'jest-localstorage-mock';
-import { addTask } from './task.js';
+import { addTask, updateTaskValue } from './task.js';
+import AddCheckBox from './completion.js';
```
Comment (adel-gu) [inline src/task.test.js:4]: - [OPTIONAL] A good project directory structure convention is to wrap up all tests files inside `__tests__` directory 👍🏼

**Review by adel-gu — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES ♻️ 
**Hi @Team 👋🏼** 

Good job so far! 👏🏼👏🏼👏🏼 however, there are some issues that you still need to work on to go to the next project but you are almost there! 💪🏼💪🏼💪🏼. 



## Highlights
- All pure functions are well tested ✅ 
- Tests are grouped using the `description()` method ✅ 
- Well done clearing the localstroge before each test suit ✅ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ismailco — APPROVED** [review]
### Hi @fredo509, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

cr-approve


## Highlights

✔️ Include DOM mock test
✔️ No linter errors
✔️ Used GitHub Flow
✔️ Work documented professionally

## Optional suggestions
- [ ] [OPTIONAL] I highly recommend adding your test files to a `_tests_` folder to keep the code base organized

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ismailco — APPROVED** [review]
### Hi @fredo509, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

cr-approve


## Highlights

✔️ Include DOM mock test
✔️ No linter errors
✔️ Used GitHub Flow
✔️ Work documented professionally

## Optional suggestions
- [x] [OPTIONAL] I highly recommend adding your test files to a `_tests_` folder to keep the code base organized

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**